go/consensus/tendermint/light/service.go

```diff
@@ -0,0 +1,337 @@
+	numProviders = 3
+	// lcMaxRetryAttempts is the number of retry attempts the tendermint light client does,
+	// before switching the primary provider.
+	lcMaxRetryAttempts = 5
+
+	// CfgConsensusLightClientTrustPeriod is the light client trust period.
+	CfgConsensusLightClientTrustPeriod = "consensus.tendermint.light_client.trust_period"
+)
+
+// Flags has the configuration flags.
+var Flags = flag.NewFlagSet("", flag.ContinueOnError)
+
+type client struct {
+	ctx context.Context
+
+	enabled bool
+
+	logger *logging.Logger
+
+	genesis   genesisAPI.Provider
+	consensus consensusAPI.Backend
+	p2p       rpc.P2P
+
+	store tmlightstore.Store
+
+	lc        *lightClient
+	providers []api.Provider
+
+	stopped uint32
+	initCh  chan struct{} // closed internally when initialized
+	stopCh  chan struct{} // closed internally to trigger stop
+	quitCh  chan struct{} // closed after stopped
+}
+
+// Name implements api.ClientService
+func (*client) Name() string {
+	return "tendermint/light"
+}
+
+func (c *client) Start() error {
+	go c.worker()
+
+	return nil
+}
+
+func (c *client) GetStatus(ctx context.Context) (*consensus.LightClientStatus, error) {
+	status := &consensus.LightClientStatus{}
+	oldest, err := c.store.FirstLightBlockHeight()
+	if err != nil {
+		return nil, err
+	}
+	status.OldestHeight = oldest
+	if oldest > -1 { // -1 indicates empty store.
+		var lb *tmtypes.LightBlock
+		lb, err = c.store.LightBlock(oldest)
+		if err != nil {
+			return nil, err
+		}
+		status.OldestHash = hash.LoadFromHexBytes(lb.Header.Hash())
+		status.OldestTime = lb.Time
+	}
+
+	latest, err := c.store.LastLightBlockHeight()
+	if err != nil {
+		return nil, err
+	}
+	status.LatestHeight = latest
+	if latest > -1 { // -1 indicates empty store.
+		var lb *tmtypes.LightBlock
+		lb, err := c.store.LightBlock(latest)
+		if err != nil {
+			return nil, err
+		}
+		status.LatestHash = hash.LoadFromHexBytes(lb.Header.Hash())
+		status.LatestTime = lb.Time
+	}
+
+	for _, p := range c.providers {
+		status.PeerIDs = append(status.PeerIDs, p.PeerID())
+	}
+
+	return status, nil
+}
+
+func (c *client) Stop() {
+	if !atomic.CompareAndSwapUint32(&c.stopped, 0, 1) {
+		return
+	}
+	close(c.stopCh)
+}
+
+func (c *client) Quit() <-chan struct{} {
+	return c.quitCh
+}
+
+func (c *client) Cleanup() {
+}
+
+func (c *client) consensusBlkToTendermint(lb *consensus.LightBlock) (*tmtypes.LightBlock, error) {
+	// Convert to protobuff LightBlock.
+	var pb tmproto.LightBlock
+	if err := pb.Unmarshal(lb.Meta); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal tendermint protobuff light block: %w", err)
+	}
+	// Convert to tendermint LightBlock.
+	tmLb, err := tmtypes.LightBlockFromProto(&pb)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert protobuff light block into tendermint light block: %w", err)
+	}
+	return tmLb, nil
+}
+
+func (c *client) worker() {
+	if !c.enabled {
+		return
+	}
+	var initOk bool
+	defer func() {
+		if !initOk {
```

Comment:
Is there a reason why this doesn't always close the quit channel, but it only does so on failures? Will this result in the service hanging when stopped regularly?

Reply:
It was wrong. I refactored the init/stop/quit channels. Should be better now.